File: server/src/routes/analyze-stream.ts
/**
 * @fileoverview Streaming URL analysis endpoint with progress updates.
 * Provides Server-Sent Events (SSE) endpoint for real-time progress during
 * page loading, consent handling, and tracking analysis.
 * 
 * Updated to use BrowserSession for concurrent request support.
 */

import type { Request, Response } from 'express'
import { BrowserSession, type DeviceType } from '../services/browser-session.js'
import { runTrackingAnalysis } from '../services/analysis.js'
import { analyzeScripts } from '../services/script-analysis.js'
import { validateOpenAIConfig } from '../services/openai.js'
import { getErrorMessage, createLogger, startLogFile } from '../utils/index.js'
import { extractDomain } from '../utils/url.js'
import { sendEvent, sendProgress, handleOverlays } from './analyze-helpers.js'

const log = createLogger('Analyze')

// ============================================================================
// Main Streaming Handler
// ============================================================================

/**
 * GET /api/open-browser-stream - Analyze tracking on a URL with streaming progress.
 *
 * Opens a browser, navigates to the URL, detects and handles cookie
 * consent banners, captures tracking data, and runs AI analysis. Progress
 * is streamed via Server-Sent Events (SSE).
 *
 * Each request creates its own isolated BrowserSession, enabling concurrent
 * analyses without interference.
 *
 * Query parameters:
 * - url: The URL to analyze (required)
 * - device: Device type to emulate (optional, default: 'ipad')
 *
 * SSE Events emitted:
 * - progress: { step, message, progress } - Progress updates
 * - screenshot: { screenshot, cookies, scripts, networkRequests, storage } - Page captures
 * - consent: { detected, clicked, details } - Consent banner handling result
 * - consentDetails: { categories, partners, purposes } - Extracted consent info
 * - complete: { success, analysis, highRisks, ... } - Final results
 * - error: { error } - If something fails
 */
export async function analyzeUrlStreamHandler(req: Request, res: Response): Promise<void> {
  const url = req.query.url as string
  const device = (req.query.device as string) || 'ipad'

  // Set up SSE headers first so we can send error events
  res.setHeader('Content-Type', 'text/event-stream')
  res.setHeader('Cache-Control', 'no-cache')
  res.setHeader('Connection', 'keep-alive')
  res.setHeader('Access-Control-Allow-Origin', '*')

  // Validate OpenAI configuration before starting analysis
  const configError = validateOpenAIConfig()
  if (configError) {
    sendEvent(res, 'error', { error: configError })
    res.end()
    return
  }

  // Validate device type
  const validDevices = ['iphone', 'ipad', 'android-phone', 'android-tablet', 'windows-chrome', 'macos-safari']
  const deviceType: DeviceType = validDevices.includes(device) ? device as DeviceType : 'ipad'

  if (!url) {
    sendEvent(res, 'error', { error: 'URL is required' })
    res.end()
    return
  }

  // Create isolated browser session for this request
  const session = new BrowserSession()

  // Start file logging for this analysis (if enabled)
  const domain = extractDomain(url)
  startLogFile(domain)

  // Start overall timer
  log.section(`Analyzing: ${url}`)
  log.info('Request received', { url, device: deviceType })
  log.startTimer('total-analysis')

  try {
    // ========================================================================
    // Phase 1: Browser Setup and Navigation
    // ========================================================================
    
    log.subsection('Phase 1: Browser Setup')
    sendProgress(res, 'init', 'Warming up...', 5)

    session.clearTrackingData()
    session.setCurrentPageUrl(url)

    log.startTimer('browser-launch')
    sendProgress(res, 'browser', 'Launching browser...', 8)
    await session.launchBrowser(deviceType)
    log.endTimer('browser-launch', 'Browser launched')

    const hostname = new URL(url).hostname
    log.startTimer('navigation')
    log.info('Navigating to page', { hostname })
    sendProgress(res, 'navigate', `Connecting to ${hostname}...`, 12)

    // Navigate and wait for DOM to be ready
    const navResult = await session.navigateTo(url, 'domcontentloaded', 30000)
    log.endTimer('navigation', 'Initial navigation complete')
    log.info('Navigation result', { success: navResult.success, statusCode: navResult.statusCode })
    
    // Check for HTTP-level errors
    if (!navResult.success) {
      const errorType = navResult.isAccessDenied ? 'access-denied' : 'server-error'
      log.error('Navigation failed', { errorType, statusCode: navResult.statusCode, message: navResult.errorMessage })
      sendEvent(res, 'pageError', {
        type: errorType,
        statusCode: navResult.statusCode,
        message: navResult.errorMessage,
        isAccessDenied: navResult.isAccessDenied,
      })
      sendProgress(res, 'error', navResult.errorMessage || 'Failed to load page', 100)
      res.end()
      return
    }
    
    // ========================================================================
    // Phase 2: Wait for Page Load and Check Access
    // ========================================================================
    
    log.subsection('Phase 2: Page Load & Access Check')
    log.startTimer('network-idle')
    sendProgress(res, 'wait-network', `Loading ${hostname}...`, 18)
    
    // Wait for network to settle (give more time for ad auctions and lazy loading)
    const networkIdleResult = await session.waitForNetworkIdle(20000)
    log.endTimer('network-idle', 'Network idle wait complete')
    
    if (!networkIdleResult) {
      log.warn('Network still active (normal for ad-heavy sites), continuing...')
      sendProgress(res, 'wait-continue', 'Page loaded, waiting for ads to render...', 25)
      // Give extra time for ad scripts to execute and render
      await session.waitForTimeout(3000)
    } else {
      log.success('Network became idle')
      sendProgress(res, 'wait-done', 'Page fully loaded', 28)
    }

    // Additional wait for lazy-loaded ads and deferred scripts
    // Ads often load via setTimeout/requestAnimationFrame after initial load
    sendProgress(res, 'wait-ads', 'Waiting for dynamic content...', 28)
    await session.waitForTimeout(2000)
    
    // Check for access denied in page content (bot detection, Cloudflare, etc.)
    log.startTimer('access-check')
    const accessCheck = await session.checkForAccessDenied()
    log.endTimer('access-check', 'Access check complete')
    
    if (accessCheck.denied) {
      log.error('Access denied detected', { reason: accessCheck.reason })
      
      const screenshot = await session.takeScreenshot(false)
      const base64Screenshot = screenshot.toString('base64')
      
      sendEvent(res, 'screenshot', {
        screenshot: `data:image/png;base64,${base64Screenshot}`,
        cookies: session.getTrackedCookies(),
        scripts: session.getTrackedScripts(),
        networkRequests: session.getTrackedNetworkRequests(),
        localStorage: [],
        sessionStorage: [],
      })
      
      sendEvent(res, 'pageError', {
        type: 'access-denied',
        statusCode: navResult.statusCode,
        message: 'Access denied - this site has bot protection that blocked our request',
        isAccessDenied: true,
        reason: accessCheck.reason,
      })
      sendProgress(res, 'blocked', 'Site blocked access', 100)
      res.end()
      return
    }

    // ========================================================================
    // Phase 3: Initial Data Capture
    // ========================================================================
    
    log.subsection('Phase 3: Initial Data Capture')
    log.startTimer('initial-capture')
    
    sendProgress(res, 'cookies', 'Capturing cookies...', 32)
    await session.captureCurrentCookies()
    
    sendProgress(res, 'storage', 'Capturing storage data...', 35)
    let storage = await session.captureStorage()

    // Wait for consent dialogs to appear
    sendProgress(res, 'overlay-wait', 'Waiting for page overlays...', 37)
    await session.waitForTimeout(2000)

    sendProgress(res, 'screenshot', 'Taking screenshot...', 38)
    const screenshot = await session.takeScreenshot(false)
    const base64Screenshot = screenshot.toString('base64')

    const cookieCount = session.getTrackedCookies().length
    const scriptCount = session.getTrackedScripts().length
    const requestCount = session.getTrackedNetworkRequests().length
    
    log.endTimer('initial-capture', 'Initial data captured')
    log.info('Initial capture stats', { cookies: cookieCount, scripts: scriptCount, requests: requestCount, localStorage: storage.localStorage.length, sessionStorage: storage.sessionStorage.length })
    
    sendProgress(res, 'captured', `Found ${cookieCount} cookies, ${scriptCount} scripts, ${requestCount} requests`, 42)

    // Send initial screenshot
    sendEvent(res, 'screenshot', {
      screenshot: `data:image/png;base64,${base64Screenshot}`,
      cookies: session.getTrackedCookies(),
      scripts: session.getTrackedScripts(),
      networkRequests: session.getTrackedNetworkRequests(),
      localStorage: storage.localStorage,
      sessionStorage: storage.sessionStorage,
    })

    // ========================================================================
    // Phase 4: Overlay Detection and Handling
    // ========================================================================
    
    log.subsection('Phase 4: Overlay Detection & Handling')
    log.startTimer('overlay-handling')
    sendProgress(res, 'overlay-detect', 'Checking for page overlays...', 45)

    const page = session.getPage()
    let consentDetails = null
    let overlayCount = 0

    if (page) {
      const overlayResult = await handleOverlays(session, page, res, screenshot)
      consentDetails = overlayResult.consentDetails
      overlayCount = overlayResult.overlayCount
      storage = overlayResult.finalStorage
    }
    
    log.endTimer('overlay-handling', 'Overlay handling complete')
    log.info('Overlay handling result', { overlaysFound: overlayCount, hasConsentDetails: !!consentDetails })

    // ========================================================================
    // Phase 5: AI Analysis
    // ========================================================================
    
    log.subsection('Phase 5: AI Analysis')
    sendProgress(res, 'analysis-prep', 'Preparing tracking data for analysis...', 75)

    const finalCookieCount = session.getTrackedCookies().length
    const finalScriptCount = session.getTrackedScripts().length
    const finalRequestCount = session.getTrackedNetworkRequests().length
    
    log.info('Final data stats', { cookies: finalCookieCount, scripts: finalScriptCount, requests: finalRequestCount })
    sendProgress(res, 'analysis-start', `Found ${finalCookieCount} cookies, ${finalScriptCount} scripts, ${finalRequestCount} requests`, 76)

    log.startTimer('ai-analysis')
    
    // Run script analysis first (grouping, pattern matching, LLM for unknowns)
    log.startTimer('script-analysis')
    const scriptAnalysisResult = await analyzeScripts(
      session.getTrackedScripts(),
      (phase, current, total, detail) => {
        if (phase === 'matching') {
          if (current === 0) {
            sendProgress(res, 'script-matching', detail || 'Grouping and identifying scripts...', 77)
          }
        } else if (phase === 'fetching') {
          // Fetching phase: 77-79%
          const progress = 77 + Math.floor((current / Math.max(total, 1)) * 2)
          sendProgress(res, 'script-fetching', detail || `Fetching script ${current}/${total}...`, progress)
        } else if (phase === 'analyzing') {
          if (total === 0) {
            sendProgress(res, 'script-analysis', detail || 'All scripts identified', 82)
          } else {
            // Analyzing phase: 79-83%
            const progress = 79 + Math.floor((current / total) * 4)
            sendProgress(res, 'script-analysis', detail || `Analyzed ${current}/${total} scripts...`, progress)
          }
        }
      }
    )
    log.endTimer('script-analysis', `Script analysis complete (${scriptAnalysisResult.scripts.length} scripts, ${scriptAnalysisResult.groups.length} groups)`)
    
    // Then run main tracking analysis
    log.startTimer('tracking-analysis')
    const analysisResult = await runTrackingAnalysis(
      session.getTrackedCookies(),
      storage.localStorage,
      storage.sessionStorage,
      session.getTrackedNetworkRequests(),
      session.getTrackedScripts(),
      url,
      consentDetails,
      (phase, detail) => {
        switch (phase) {
          case 'preparing':
            sendProgress(res, 'ai-preparing', detail || 'Preparing data for AI analysis...', 84)
            break
          case 'analyzing':
            sendProgress(res, 'ai-analyzing', detail || 'Generating privacy report...', 88)
            break
          case 'scoring':
            sendProgress(res, 'ai-scoring', detail || 'Calculating privacy score...', 94)
            break
          case 'summarizing':
            sendProgress(res, 'ai-summarizing', detail || 'Generating summary findings...', 97)
            break
        }
      }
    )
    log.endTimer('tracking-analysis', 'Tracking analysis complete')
    
    log.endTimer('ai-analysis', 'All AI analysis complete')

    if (analysisResult.success) {
      log.success('Analysis succeeded', { privacyScore: analysisResult.privacyScore, analysisLength: analysisResult.analysis?.length })
    } else {
      log.error('Analysis failed', { error: analysisResult.error })
    }

    // ========================================================================
    // Phase 6: Complete
    // ========================================================================
    
    const totalTime = log.endTimer('total-analysis', 'Analysis complete')
    log.success('Investigation complete!', { totalTime: `${(totalTime / 1000).toFixed(2)}s`, overlaysDismissed: overlayCount, privacyScore: analysisResult.privacyScore })
    
    sendProgress(res, 'complete', 'Investigation complete!', 100)

    sendEvent(res, 'complete', {
      success: true,
      message: overlayCount > 0
        ? 'Tracking analyzed after dismissing overlays'
        : 'Tracking analyzed',
      analysis: analysisResult.success ? analysisResult.analysis : null,
      summaryFindings: analysisResult.success ? analysisResult.summaryFindings : null,
      privacyScore: analysisResult.success ? analysisResult.privacyScore : null,
      privacySummary: analysisResult.success ? analysisResult.privacySummary : null,
      analysisSummary: analysisResult.success ? analysisResult.summary : null,
      analysisError: analysisResult.success ? null : analysisResult.error,
      consentDetails: consentDetails,
      scripts: scriptAnalysisResult.scripts,
      scriptGroups: scriptAnalysisResult.groups,
    })

    res.end()
  } catch (error) {
    log.error('Analysis failed with exception', { error: getErrorMessage(error) })
    sendEvent(res, 'error', { error: getErrorMessage(error) })
    res.end()
  } finally {
    // Always clean up browser resources to prevent memory leaks
    log.debug('Cleaning up browser resources...')
    await session.close().catch((err) => {
      log.warn('Error during browser cleanup', { error: getErrorMessage(err) })
    })
    log.debug('Browser cleanup complete')
  }
}

File: server/src/routes/analyze_stream.py
"""
Streaming URL analysis endpoint with progress updates.
Provides Server-Sent Events (SSE) endpoint for real-time progress during
page loading, consent handling, and tracking analysis.
"""

from __future__ import annotations

import asyncio
from typing import Any, AsyncGenerator, cast
from urllib import parse

from src.routes import analyze_helpers
from src.agents import config as agent_config
from src.services import (
    analysis,
    browser_session,
    device_configs,
    script_analysis,
)
from src.types import browser
from src.utils import errors, logger
from src.utils import url as url_mod

log = logger.create_logger("Analyze")


async def analyze_url_stream(url: str, device: str = "ipad") -> AsyncGenerator[str, None]:
    """
    Analyze tracking on a URL with streaming progress via SSE.

    Opens a browser, navigates to the URL, detects and handles cookie
    consent banners, captures tracking data, and runs AI analysis. Progress
    is streamed via Server-Sent Events (SSE).

    Each call creates its own isolated BrowserSession, enabling concurrent
    analyses without interference.
    """
    # Validate LLM configuration
    config_error = agent_config.validate_llm_config()
    if config_error:
        yield analyze_helpers.format_sse_event("error", {"error": config_error})
        return

    # Validate device type
    valid_devices = list(device_configs.DEVICE_CONFIGS.keys())
    device_type = cast(browser.DeviceType, device if device in valid_devices else "ipad")

    if not url:
        yield analyze_helpers.format_sse_event("error", {"error": "URL is required"})
        return

    # Create isolated browser session
    session = browser_session.BrowserSession()

    # Start file logging
    domain = url_mod.extract_domain(url)
    logger.start_log_file(domain)

    log.section(f"Analyzing: {url}")
    log.info("Request received", {"url": url, "device": device_type})
    log.start_timer("total-analysis")

    try:
        # ====================================================================
        # Phase 1: Browser Setup and Navigation
        # ====================================================================
        log.subsection("Phase 1: Browser Setup")
        yield analyze_helpers.format_progress_event("init", "Warming up...", 5)

        session.clear_tracking_data()
        session.set_current_page_url(url)

        log.start_timer("browser-launch")
        yield analyze_helpers.format_progress_event("browser", "Launching browser...", 8)
        await session.launch_browser(device_type)
        log.end_timer("browser-launch", "Browser launched")

        hostname = parse.urlparse(url).hostname or url
        log.start_timer("navigation")
        log.info("Navigating to page", {"hostname": hostname})
        yield analyze_helpers.format_progress_event("navigate", f"Connecting to {hostname}...", 12)

        nav_result = await session.navigate_to(url, "domcontentloaded", 30000)
        log.end_timer("navigation", "Initial navigation complete")
        log.info("Navigation result", {"success": nav_result.success, "statusCode": nav_result.status_code})

        if not nav_result.success:
            error_type = "access-denied" if nav_result.is_access_denied else "server-error"
            log.error("Navigation failed", {"errorType": error_type, "statusCode": nav_result.status_code})
            yield analyze_helpers.format_sse_event("pageError", {
                "type": error_type,
                "statusCode": nav_result.status_code,
                "message": nav_result.error_message,
                "isAccessDenied": nav_result.is_access_denied,
            })
            yield analyze_helpers.format_progress_event("error", nav_result.error_message or "Failed to load page", 100)
            return

        # ====================================================================
        # Phase 2: Wait for Page Load and Check Access
        # ====================================================================
        log.subsection("Phase 2: Page Load & Access Check")
        log.start_timer("network-idle")
        yield analyze_helpers.format_progress_event("wait-network", f"Loading {hostname}...", 18)

        network_idle = await session.wait_for_network_idle(20000)
        log.end_timer("network-idle", "Network idle wait complete")

        if not network_idle:
            log.warn("Network still active (normal for ad-heavy sites), continuing...")
            yield analyze_helpers.format_progress_event("wait-continue", "Page loaded, waiting for ads to render...", 25)
            await session.wait_for_timeout(3000)
        else:
            log.success("Network became idle")
            yield analyze_helpers.format_progress_event("wait-done", "Page fully loaded", 28)

        yield analyze_helpers.format_progress_event("wait-ads", "Waiting for dynamic content...", 28)
        await session.wait_for_timeout(2000)

        # Check for access denied
        log.start_timer("access-check")
        access_check = await session.check_for_access_denied()
        log.end_timer("access-check", "Access check complete")

        if access_check.denied:
            log.error("Access denied detected", {"reason": access_check.reason})
            screenshot = await session.take_screenshot(full_page=False)
            optimized = browser_session.BrowserSession.optimize_screenshot_bytes(screenshot)

            yield analyze_helpers.format_sse_event("screenshot", {
                "screenshot": optimized,
                "cookies": [analyze_helpers.to_camel_case_dict(c) for c in session.get_tracked_cookies()],
                "scripts": [analyze_helpers.to_camel_case_dict(s) for s in session.get_tracked_scripts()],
                "networkRequests": [analyze_helpers.to_camel_case_dict(r) for r in session.get_tracked_network_requests()],
                "localStorage": [],
                "sessionStorage": [],
            })
            yield analyze_helpers.format_sse_event("pageError", {
                "type": "access-denied",
                "statusCode": nav_result.status_code,
                "message": "Access denied - this site has bot protection that blocked our request",
                "isAccessDenied": True,
                "reason": access_check.reason,
            })
            yield analyze_helpers.format_progress_event("blocked", "Site blocked access", 100)
            return

        # ====================================================================
        # Phase 3: Initial Data Capture
        # ====================================================================
        log.subsection("Phase 3: Initial Data Capture")
        log.start_timer("initial-capture")

        yield analyze_helpers.format_progress_event("cookies", "Capturing cookies...", 32)
        await session.capture_current_cookies()

        yield analyze_helpers.format_progress_event("storage", "Capturing storage data...", 35)
        storage = await session.capture_storage()

        yield analyze_helpers.format_progress_event("overlay-wait", "Waiting for page overlays...", 37)
        await session.wait_for_timeout(2000)

        yield analyze_helpers.format_progress_event("screenshot", "Taking screenshot...", 38)
        screenshot = await session.take_screenshot(full_page=False)
        optimized_screenshot = browser_session.BrowserSession.optimize_screenshot_bytes(screenshot)

        # Send screenshot to client immediately
        yield analyze_helpers.format_sse_event("screenshot", {
            "screenshot": optimized_screenshot,
            "cookies": [analyze_helpers.to_camel_case_dict(c) for c in session.get_tracked_cookies()],
            "scripts": [analyze_helpers.to_camel_case_dict(s) for s in session.get_tracked_scripts()],
            "networkRequests": [analyze_helpers.to_camel_case_dict(r) for r in session.get_tracked_network_requests()],
            "localStorage": [analyze_helpers.to_camel_case_dict(i) for i in storage["local_storage"]],
            "sessionStorage": [analyze_helpers.to_camel_case_dict(i) for i in storage["session_storage"]],
        })

        cookie_count = len(session.get_tracked_cookies())
        script_count = len(session.get_tracked_scripts())
        request_count = len(session.get_tracked_network_requests())

        log.end_timer("initial-capture", "Initial data captured")
        log.info("Initial capture stats", {
            "cookies": cookie_count,
            "scripts": script_count,
            "requests": request_count,
            "localStorage": len(storage["local_storage"]),
            "sessionStorage": len(storage["session_storage"]),
        })

        yield analyze_helpers.format_progress_event(
            "captured",
            f"Found {cookie_count} cookies, {script_count} scripts, {request_count} requests",
            42,
        )

        # ====================================================================
        # Phase 4: Overlay Detection and Handling
        # ====================================================================
        log.subsection("Phase 4: Overlay Detection & Handling")
        log.start_timer("overlay-handling")
        yield analyze_helpers.format_progress_event("overlay-detect", "Checking for page overlays...", 45)

        page = session.get_page()
        consent_details = None
        overlay_count = 0

        if page:
            overlay_result = analyze_helpers.OverlayHandlingResult()
            async for event_str in analyze_helpers.handle_overlays(session, page, screenshot, overlay_result):
                yield event_str
            consent_details = overlay_result.consent_details
            overlay_count = overlay_result.overlay_count
            storage = overlay_result.final_storage

        log.end_timer("overlay-handling", "Overlay handling complete")
        log.info("Overlay handling result", {"overlaysFound": overlay_count, "hasConsentDetails": consent_details is not None})

        # Capture a fresh screenshot after overlay handling.
        yield analyze_helpers.format_progress_event("post-overlay-screenshot", "Capturing final page state...", 72)
        screenshot = await session.take_screenshot(full_page=False)
        optimized_screenshot = browser_session.BrowserSession.optimize_screenshot_bytes(screenshot)
        await session.capture_current_cookies()
        storage = await session.capture_storage()

        yield analyze_helpers.format_sse_event("screenshot", {
            "screenshot": optimized_screenshot,
            "cookies": [analyze_helpers.to_camel_case_dict(c) for c in session.get_tracked_cookies()],
            "scripts": [analyze_helpers.to_camel_case_dict(s) for s in session.get_tracked_scripts()],
            "networkRequests": [analyze_helpers.to_camel_case_dict(r) for r in session.get_tracked_network_requests()],
            "localStorage": [analyze_helpers.to_camel_case_dict(i) for i in storage["local_storage"]],
            "sessionStorage": [analyze_helpers.to_camel_case_dict(i) for i in storage["session_storage"]],
        })

        # Abort if an overlay could not be dismissed — the page is blocked.
        if page and overlay_result.failed:
            log.error("Overlay dismissal failed, aborting analysis", {"reason": overlay_result.failure_message})
            yield analyze_helpers.format_sse_event("pageError", {
                "type": "overlay-blocked",
                "message": overlay_result.failure_message,
                "isOverlayBlocked": True,
            })
            yield analyze_helpers.format_progress_event("overlay-blocked", "Could not dismiss page overlay", 100)
            return

        # ====================================================================
        # Phase 5: AI Analysis
        # ====================================================================
        log.subsection("Phase 5: AI Analysis")
        yield analyze_helpers.format_progress_event("analysis-prep", "Preparing tracking data for analysis...", 75)

        final_cookies = session.get_tracked_cookies()
        final_scripts = session.get_tracked_scripts()
        final_requests = session.get_tracked_network_requests()

        log.info("Final data stats", {
            "cookies": len(final_cookies),
            "scripts": len(final_scripts),
            "requests": len(final_requests),
        })
        yield analyze_helpers.format_progress_event(
            "analysis-start",
            f"Found {len(final_cookies)} cookies, {len(final_scripts)} scripts, {len(final_requests)} requests",
            76,
        )

        log.start_timer("ai-analysis")

        # ---- Shared progress queue for concurrent tasks ----
        progress_queue: asyncio.Queue[str | None] = asyncio.Queue()
        _tasks_remaining = 2  # script analysis + streaming analysis

        # ---- Script analysis (runs concurrently) ----
        log.start_timer("script-analysis")

        def script_progress(phase: str, current: int, total: int, detail: str) -> None:
            """Push progress events onto the async queue."""
            log.info(f"Script analysis progress: {phase} {current}/{total} - {detail}")
            if phase == "matching":
                if current == 0:
                    progress_queue.put_nowait(
                        analyze_helpers.format_progress_event("script-matching", detail or "Grouping and identifying scripts...", 77)
                    )
            elif phase == "fetching":
                progress = 77 + int((current / max(total, 1)) * 2)
                progress_queue.put_nowait(
                    analyze_helpers.format_progress_event("script-fetching", detail or f"Fetching script {current}/{total}...", progress)
                )
            elif phase == "analyzing":
                if total == 0:
                    progress_queue.put_nowait(
                        analyze_helpers.format_progress_event("script-analysis", detail or "All scripts identified", 82)
                    )
                else:
                    progress = 79 + int((current / total) * 11)
                    progress_queue.put_nowait(
                        analyze_helpers.format_progress_event("script-analysis", detail or f"Analyzed {current}/{total} scripts...", progress)
                    )

        async def run_script_analysis() -> Any:
            """Run script analysis and signal the queue."""
            try:
                result = await script_analysis.analyze_scripts(final_scripts, script_progress)
                log.end_timer("script-analysis", f"Script analysis complete ({len(result.scripts)} scripts, {len(result.groups)} groups)")
                return result
            finally:
                progress_queue.put_nowait(None)

        # ---- Streaming tracking analysis (runs concurrently) ----
        log.start_timer("tracking-analysis")
        analysis_chunks: list[str] = []

        async def run_streaming_analysis() -> None:
            """Stream tracking analysis and push chunks + progress to queue."""
            try:
                async for chunk in analysis.stream_tracking_analysis(
                    final_cookies,
                    storage["local_storage"],
                    storage["session_storage"],
                    final_requests,
                    final_scripts,
                    url,
                    consent_details,
                ):
                    analysis_chunks.append(chunk)
                    progress_queue.put_nowait(
                        analyze_helpers.format_sse_event("analysis-chunk", {"text": chunk})
                    )
            finally:
                progress_queue.put_nowait(None)

        # Launch both concurrently
        script_task = asyncio.create_task(run_script_analysis())
        analysis_stream_task = asyncio.create_task(run_streaming_analysis())

        # Drain progress events until both signal completion
        finished = 0
        while finished < _tasks_remaining:
            event = await progress_queue.get()
            if event is None:
                finished += 1
                continue
            yield event

        await script_task
        await analysis_stream_task
        script_analysis_result = script_task.result()

        log.end_timer("tracking-analysis", "Tracking analysis complete")

        # ---- Post-streaming: scoring + summary findings ----
        full_analysis_text = "".join(analysis_chunks)
        log.info("Analysis streamed", {"length": len(full_analysis_text)})

        yield analyze_helpers.format_progress_event("ai-scoring", "Calculating privacy score...", 94)
        from src.services import privacy_score as privacy_score_mod
        from src.utils import tracking_summary as tracking_summary_mod

        tracking_summary = tracking_summary_mod.build_tracking_summary(
            final_cookies, final_scripts, final_requests,
            storage["local_storage"], storage["session_storage"], url,
        )

        score_breakdown = privacy_score_mod.calculate_privacy_score(
            final_cookies, final_scripts, final_requests,
            storage["local_storage"], storage["session_storage"],
            url, consent_details,
        )

        yield analyze_helpers.format_progress_event("ai-summarizing", "Generating summary findings...", 97)
        from src.agents import get_summary_findings_agent
        summary_agent = get_summary_findings_agent()
        summary_findings = await summary_agent.summarise(full_analysis_text)

        log.end_timer("ai-analysis", "All AI analysis complete")

        analysis_success = bool(full_analysis_text)
        privacy_score = score_breakdown.total_score
        privacy_summary = score_breakdown.summary

        if analysis_success:
            log.success("Analysis succeeded", {
                "privacyScore": privacy_score,
                "analysisLength": len(full_analysis_text),
            })
        else:
            log.error("Analysis produced no output")

        # ====================================================================
        # Phase 6: Complete
        # ====================================================================
        total_time = log.end_timer("total-analysis", "Analysis complete")
        log.success("Investigation complete!", {
            "totalTime": f"{(total_time / 1000):.2f}s",
            "overlaysDismissed": overlay_count,
            "privacyScore": privacy_score,
        })

        yield analyze_helpers.format_progress_event("complete", "Investigation complete!", 100)

        # Build consent details for response
        consent_details_dict = analyze_helpers.serialize_consent_details(consent_details) if consent_details else None

        # Build score breakdown dict
        score_breakdown_dict = (
            analyze_helpers.serialize_score_breakdown(score_breakdown)
            if score_breakdown
            else None
        )

        yield analyze_helpers.format_sse_event("complete", {
            "success": True,
            "message": (
                "Tracking analyzed after dismissing overlays"
                if overlay_count > 0
                else "Tracking analyzed"
            ),
            "analysis": full_analysis_text if analysis_success else None,
            "summaryFindings": (
                [{"type": f.type, "text": f.text} for f in summary_findings]
                if analysis_success
                else None
            ),
            "privacyScore": privacy_score if analysis_success else None,
            "privacySummary": privacy_summary if analysis_success else None,
            "scoreBreakdown": score_breakdown_dict if analysis_success else None,
            "analysisSummary": (
                {
                    "analyzedUrl": tracking_summary.analyzed_url,
                    "totalCookies": tracking_summary.total_cookies,
                    "totalScripts": tracking_summary.total_scripts,
                    "totalNetworkRequests": tracking_summary.total_network_requests,
                    "localStorageItems": tracking_summary.local_storage_items,
                    "sessionStorageItems": tracking_summary.session_storage_items,
                    "thirdPartyDomains": tracking_summary.third_party_domains,
                    "domainBreakdown": [analyze_helpers.to_camel_case_dict(d) for d in tracking_summary.domain_breakdown],
                    "localStorage": tracking_summary.local_storage,
                    "sessionStorage": tracking_summary.session_storage,
                }
                if tracking_summary
                else None
            ),
            "analysisError": None if analysis_success else "Analysis produced no output",
            "consentDetails": consent_details_dict,
            "scripts": [analyze_helpers.to_camel_case_dict(s) for s in script_analysis_result.scripts],
            "scriptGroups": [analyze_helpers.to_camel_case_dict(g) for g in script_analysis_result.groups],
        })

    except Exception as error:
        log.error("Analysis failed with exception", {"error": errors.get_error_message(error)})
        yield analyze_helpers.format_sse_event("error", {"error": errors.get_error_message(error)})
    finally:
        log.debug("Cleaning up browser resources...")
        try:
            await session.close()
        except Exception as err:
            log.warn("Error during browser cleanup", {"error": errors.get_error_message(err)})
        log.debug("Browser cleanup complete")
        logger.end_log_file()
